### DISCUSSION TITLE
Apple MLX framework released

### DISCUSSION THREAD

**Discussion by AdithyanI** [ggml-org/llama.cpp#4345] [Ideas]
I just stumbled upon this : https://github.com/ml-explore/mlx

"MLX is an array framework for machine learning on Apple silicon, brought to you by Apple machine learning research."

<img width="802" alt="image" src="https://github.com/ggerganov/llama.cpp/assets/12037164/b080de1b-6d2a-490a-9001-756328f5a33e">

Can someone help me understand, how will this affect llama.cpp and whisper.cpp? 
Looks like in the examples they quote those. 

Can we leverage this in our repos and make them even faster? 

Best,
Adi

**Comment by slaren** [comment]
From what I see, this seems to be like Apple's equivalent of pytorch, and it is too high level for what we need in ggml. However, the source code has a Metal backend, and we may be able to use it to learn how to better optimize our Metal kernels.

**Reply by AdithyanI** [reply]
Thanks for the reply. That helps clarify some things.

**Reply by sandangel** [reply]
I'm also interested in this. Looks forward to seeing if we can reuse the code and optimize the performance further.

**Comment by 0xtotem** [comment]
Performance wise I noticed [here](https://ml-explore.github.io/mlx/build/html/examples/llama-inference.html#weight-loading-and-benchmarking):

> We observe that 4.3 seconds are required to generate 100 tokens and 0.4 seconds of those are spent processing the prompt. This amounts to a little over 39 ms per token.

This was run on a M1 Ultra and the 7B parameter Llama model (I assume Llama 2).

According to [llama.cpp's benchmark](https://github.com/ggerganov/llama.cpp/discussions/4167#user-content-fn-1-25923a8d22ca45f943c40370e25e7ceb) for the M1 Ultra 48 GPUs, we have 13.35ms/t (74.93t/s) for the Q4_0 TG.

I don't see any mention of quantization in their tutorial. So 39ms/t unquantized vs 13ms/t Q4 (assuming the same M1 with 48GPUs).

**Reply by signalprime** [reply]
The [mlx examples](https://github.com/ml-explore/mlx-examples) have been updated to include quantization examples and the recently launched [mlx-community](https://huggingface.co/mlx-community) on Hugging Face already has pre-quantized versions of llama2 and mistral available for download.  More impressively, [LLM in a flash](https://arxiv.org/pdf/2312.11514.pdf) documents context-adaptive loading to allow you to load LLMs using memory equaling roughly half the size of the model when stored on disk... in other words we'll soon be loading much bigger models without upgrading our hardware.

**Reply by AndreasKunar** [reply]
MLX this week released a version which now supports quantization .  I did a benchmarking comparison of their llama inference example against llama.cpp with llama-2 7B in Q4 and fp16 (if anyone wants to replicate/test see my github for a tweaked llama.py of theirs with token/s measures (called llama-perf.py in my repo). I'm using plain llama.cpp code, not the perf-measurement example for benchmarking.

a) MLX model load-times are very slow, GGUF and loading really rocks!
b) at 1st glance MLX prompt-processing of llama vs. the same prompt in llama.cpp seems faster. Need to dig into this
c) MLX token-generation performance for fp16 seems quite similar to llama.cpp, but 2x slower for Q4
d) memory-consumption seems similar

**Reply by WiseFarAI** [reply]
I have also followed this with great interest. I cannot wait for these techniques to be integrated with llama.cpp

**Comment by AndreasKunar** [comment]
FYI, MLX v.0.0.9 just also added experimental GGUF file support (https://github.com/ml-explore/mlx/pull/350)

**Comment by ianscrivener** [comment]
Now 6 months on from the release of MLX... I'm curious to ask to know if MLX has been beneficial to llama.cpp?

**Comment by ciekawy** [comment]
in particular what about combining both MLX and MPS?

**Comment by redthing1** [comment]
locallama users report performance of MLX vs Metal compute:
- https://old.reddit.com/r/LocalLLaMA/comments/1fz6z79/lm_studio_ships_an_mlx_backend_run_any_llm_from/
- https://old.reddit.com/r/LocalLLaMA/comments/1dazwoo/llamacpp_vs_mlx_on_mac/

Main takeaways: seems like MLX allows for dramatically reduced VRAM usage. Could allow us to squeeze larger models onto our macs.

**Comment by ejrydhfs** [comment]
I wonder if now parts of MLX could be merged into llama.cpp for everyone to use
